### PR TITLE
Smarthint Integration

### PR DESCRIPTION
**Pull request by IncognitaDev**
## What is this contribution about?

This PR add smarthint loaders, actions and tracking

## Link
https://deco-sites-prohall--smarthint.deno.dev/

### CODE REVIEW

**Review by gabrielMatosBoubee — CHANGES_REQUESTED** [review]

File: smarthint/actions/pageview.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    date,
+    origin: url.origin,
+    shcode,
+    session,
+    pageType,
+    elapsedTime: String(elapsedTime),
+    productPrice: String(productPrice),
+    shippingTime: String(shippingTime),
+    shippingPrice: String(shippingPrice),
+    anonymousConsumer: "1", //TODO
```
[the remark this answers is not in the document]
Comment (IncognitaDev) [reply smarthint/actions/pageview.ts:52]: Sure! i think this just passed by my eyes, i use the correct way before in this component

File: smarthint/sections/Analytics/SmarthintTracking.tsx [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  function getCookie(name: string): string | null {
+    const nameEQ = name + "=";
+    const ca = document.cookie.split(";");
+    for (let i = 0; i < ca.length; i++) {
+      let c = ca[i];
+      while (c.charAt(0) === " ") c = c.substring(1, c.length);
+      if (c.indexOf(nameEQ) === 0) return c.substring(nameEQ.length, c.length);
+    }
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (IncognitaDev) [reply smarthint/sections/Analytics/SmarthintTracking.tsx:82]: Just a helper function but i will refact

**Review by gabrielMatosBoubee — APPROVED** [review]
LGTM

File: smarthint/actions/click.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * @title Smarthint Integration
+ * @description Pageview Click Event
+ */
+const action = async (
+  props: Props,
+  req: Request,
+  ctx: AppContext,
+): Promise<null> => {
```
Comment (mcandeia) [inline smarthint/actions/click.ts:31]: why not void?

File: smarthint/actions/click.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    position: String(position),
+    productId: productGroupID,
+    productPrice: String(productPrice),
+    shippingTime: String(shippingTime),
+    shippingPrice: String(shippingPrice),
+    anonymousConsumer: anonymous,
+    clickProduct,
+    clickFeature,
+    locationRecs: positionRecommendation,
+  }).then((r) => r.json());
```
Comment (mcandeia) [inline smarthint/actions/click.ts:66]: why converting to json if result is not being used? json parse would unnecessary increase CPU usage in this case

File: smarthint/actions/pageview.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * @title Smarthint Integration
+ * @description Pageview Event
+ */
+const action = async (
+  props: Props,
+  req: Request,
+  ctx: AppContext,
+): Promise<null> => {
```
Comment (mcandeia) [inline smarthint/actions/pageview.ts:25]: void instead?

File: smarthint/actions/pageview.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    origin: url.origin,
+    shcode,
+    session,
+    pageType,
+    elapsedTime: String(elapsedTime),
+    productPrice: String(productPrice),
+    shippingTime: String(shippingTime),
+    shippingPrice: String(shippingPrice),
+    anonymousConsumer: anonymous,
+  }).then((r) => r.json());
```
Comment (mcandeia) [inline smarthint/actions/pageview.ts:51]: do not parse if it's not being used

File: smarthint/loaders/PLPBanners.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
+import { AppContext } from "../mod.ts";
+import { Banner, FilterProp, SearchSort } from "../utils/typings.ts";
+import { getSessionCookie } from "../utils/getSession.ts";
+import { getFilterParam, getSortParam } from "../utils/transform.ts";
+import { ProductListingPage } from "../../commerce/types.ts";
+import { getCategoriesParam } from "./recommendations.ts";
+
+export type RuleType = "valuedouble" | "valuedate" | "valuestring";
```
Comment (mcandeia) [inline smarthint/loaders/PLPBanners.ts:8]: interesting, is the word `value` needed?

File: smarthint/loaders/productListingPage.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+   */
+  ruletype?: RuleType;
+  /**
+   * @hide
+   */
+  condition?: {
+    field?: string;
+    value?: string;
+    validation?: string;
+  };
```
Comment (mcandeia) [inline smarthint/loaders/productListingPage.ts:54]: can you add a JS comment explaining why these properties are hidden?

File: smarthint/loaders/productListingPage.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+ * @description Product List Page
+ */
+const loader = async (
+  props: Props,
+  req: Request,
+  ctx: AppContext,
+): Promise<ProductListingPage | null> => {
+  const { api, shcode, cluster } = ctx;
+  const {
+    size = 12,
```
Comment (mcandeia) [inline smarthint/loaders/productListingPage.ts:72]: shouldn't this be configurable?

File: smarthint/utils/temp.getSuggestionTerms.json [hunk 10]
```diff
@@ -0,0 +1,261 @@
+{
```
Comment (mcandeia) [inline smarthint/utils/temp.getSuggestionTerms.json:1]: is this necessary?

File: smarthint/utils/temp.banners.json [hunk 11]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (mcandeia) [inline smarthint/utils/temp.banners.json:1]: is this necessary?

**Review by mcandeia — CHANGES_REQUESTED** [review]
remove `temp.*` files if they are not necessary

File: smarthint/loaders/PLPBanners.ts (hunk 7 above)
Comment (IncognitaDev) [reply smarthint/loaders/PLPBanners.ts:8]: Yes. https://developers.smarthint.co/#ea67ec69-4c0c-4004-8b6d-50f1c016d2e9

File: smarthint/loaders/productListingPage.ts (hunk 9 above)
Comment (IncognitaDev) [reply smarthint/loaders/productListingPage.ts:72]: `size` came from props, and `shcode/cluster` came from app config

**Review by mcandeia — APPROVED** [review]

File: smarthint/loaders/productListingPage.ts (hunk 8 above)
Comment (IncognitaDev) [reply smarthint/loaders/productListingPage.ts:54]: SH colaborators ask me to remove this props from admin and this props should be generated by the loaders itself. By receiving these props, we can use the loader inline for infinite scroll or other functionalities.

Comment (guitavano) [reply smarthint/loaders/productListingPage.ts:54]: I think what Candeia said is to add the comment at the code.

**Pull request merged**